File: agbac-min/agents/java/InSessionApiCall.java
package com.agbac.api;

import com.agbac.token.InSessionTokenRequest;

import java.util.Map;

/**
 * Represents an API/resource call from an in-session agent.
 * Requires a valid OAuth token obtained from InSessionTokenRequest.
 */
public class InSessionApiCall {

    private final InSessionTokenRequest tokenRequest;

    public InSessionApiCall(InSessionTokenRequest tokenRequest) {
        this.tokenRequest = tokenRequest;
    }

    /**
     * Calls a protected resource with the agent's token and human act info.
     *
     * @param sessionData Authenticated user session
     * @param agentClientId Agent client ID
     * @param resourceUrl URL of the API/resource
     * @return response body string
     */
    public String callResource(Map<String, Object> sessionData, String agentClientId, String resourceUrl) {
        try {
            String accessToken = tokenRequest.requestToken(sessionData, agentClientId);

            // Example: call API using HTTP client (pseudo-code)
            // HttpRequest req = HttpRequest.newBuilder()
            //     .uri(URI.create(resourceUrl))
            //     .header("Authorization", "Bearer " + accessToken)
            //     .GET()
            //     .build();
            // HttpResponse<String> resp = HttpClient.newHttpClient().send(req, BodyHandlers.ofString());

            String response = "Simulated API response for resource: " + resourceUrl;
            log("API called successfully with access token: " + accessToken);
            return response;

        } catch (Exception e) {
            log("Failed API/resource call: " + e.getMessage());
            throw new RuntimeException("In-session API/resource call failed", e);
        }
    }

    private void log(String message) {
        System.out.println("[InSessionApiCall] " + message);
    }
}

File: agbac-min/agents/typescript/insession/inSessionApiCall.ts
/**
 * inSessionApiCall.ts
 * Performs API or resource call for in-session agent using the token obtained with `act` claim.
 */

import axios from 'axios';
import winston from 'winston';

const logger = winston.createLogger({
    level: 'info',
    format: winston.format.combine(
        winston.format.timestamp(),
        winston.format.printf(({ timestamp, level, message }) => `${timestamp} [${level}]: ${message}`)
    ),
    transports: [new winston.transports.Console()]
});

export class InSessionApiCall {
    private apiUrl: string;

    constructor(apiUrl: string) {
        this.apiUrl = apiUrl;
    }

    /**
     * Call the API or resource using the provided token
     * @param token OAuth access token including `act` claim
     */
    async callApi(token: string): Promise<any> {
        try {
            const response = await axios.get(this.apiUrl, {
                headers: {
                    'Authorization': `Bearer ${token}`,
                    'Content-Type': 'application/json'
                }
            });

            logger.info(`API call successful. Status: ${response.status}`);
            return response.data;

        } catch (error) {
            logger.error(`API call failed: ${error.message || error}`);
            throw error;
        }
    }
}
